client/parser/test_helper.go

```diff
@@ -1,51 +1,83 @@
 package parser
 
 import (
+	"math"
 	"testing"
 
 	"alda.io/client/model"
 	"github.com/go-test/deep"
 )
 
-// A parseTestCase models the score updates that should result from parsing a
-// string of Alda code.
+// parseTestCase models a test of the `parser` module
+// parseTestCase also tests Score Updating which tests the `model` module
+// Perhaps this should live in another package that consumes both `parser` and
+// `model`, but for now we keep it here
 type parseTestCase struct {
-	label  string
-	given  string
-	expect []model.ScoreUpdate
+	label            string
+	given            string
+	expect           []model.ScoreUpdate // optional
+	expectAST        *ASTNode            // optional
+	scoreApplyOptOut bool                // optional
 }
 
 // executeParseTestCases parses each test case's given string of Alda code and
-// compares the result with the expected sequence of score updates.
+// tests the parser, scanner, ASTNode.Updates, GenerateASTFromScoreUpdates, and
+// Score Updating (with the ability to opt out)
 func executeParseTestCases(t *testing.T, testCases ...parseTestCase) {
 	for _, testCase := range testCases {
-		// We're suppressing the source context here because we're about to do a
-		// deep diff of our expected list of score updates (which are all devoid of
-		// source context) and the actual list of score updates that result from
-		// parsing the score, and we need them to be considered the same even if the
-		// source context differs.
-		//
-		// By default, the actual list of updates will include source context, which
-		// would cause an avalanche of spurious test failures because the diff would
-		// show numerous differences in the source contexts.
-		actualAST, err :=
-			Parse(testCase.label, testCase.given, SuppressSourceContext)
+		deep.MaxDepth = math.MaxInt32
+
+		// We suppress source context to facilitate deep diff comparison
+		actualAST, err := Parse(
+			testCase.label, testCase.given, SuppressSourceContext)
 		if err != nil {
 			t.Errorf("%v\n", err)
 			return
 		}
+		if testCase.expectAST != nil {
+			diff := deep.Equal(testCase.expectAST, actualAST);
+			if diff != nil {
+				t.Error(testCase.label)
+				for _, diffItem := range diff {
+					t.Errorf("%v", diffItem)
+				}
+			}
+		}
 
 		actualUpdates, err := actualAST.Updates()
 		if err != nil {
 			t.Errorf("%v\n", err)
 			return
 		}
+		if testCase.expect != nil {
+			diff := deep.Equal(testCase.expect, actualUpdates)
+			if diff != nil {
+				t.Error(testCase.label)
+				for _, diffItem := range diff {
+					t.Errorf("%v", diffItem)
+				}
+			}
+		}
 
-		if diff := deep.Equal(testCase.expect, actualUpdates); diff != nil {
+		generatedAST, err := GenerateASTFromScoreUpdates(actualUpdates)
+		if err != nil {
+			t.Errorf("%v\n", err)
+			return
+		}
+		if diff := deep.Equal(actualAST, generatedAST); diff != nil {
```

Comment:
This is great! I think it might be good to add a comment to make our intent clear: we want to ensure that the "round trip" where we translate an AST into ScoreUpdates, and then back to an AST, results in an AST equivalent to the original one.

Reply:
Good comment to add. Done.